pkg/controllers/at.go

```diff
@@ -0,0 +1,58 @@
+/*
+ (c) Copyright [2021] Micro Focus or one of its affiliates.
+ Licensed under the Apache License, Version 2.0 (the "License");
+ You may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	vapi "github.com/vertica/vertica-kubernetes/api/v1beta1"
+	"github.com/vertica/vertica-kubernetes/pkg/cmds"
+	"github.com/vertica/vertica-kubernetes/pkg/names"
+	"github.com/vertica/vertica-kubernetes/pkg/paths"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// distributeAdmintoolsConf will copy an admintools.conf to all of the running pods
+func distributeAdmintoolsConf(ctx context.Context, pf *PodFacts, pr cmds.PodRunner, atConfTempFile string) error {
+	for _, p := range pf.Detail {
+		if !p.isPodRunning {
+			continue
+		}
+		_, _, err := pr.CopyToPod(ctx, p.name, names.ServerContainer, atConfTempFile, paths.AdminToolsConf)
```

Comment:
Is it possible that the copying succeed on some Pods but fail on other Pods?

Reply:
Yes it is possible.  We make sure that all pods are running before we begin install/uninstall, but that just shortens the timing.  Do you think it is necessary to have a step in the operator to ensure admintools.conf is the same on all of the pods?  And copy if they are different.  The same pod is always picked first, so it should be a well known location that has the correct version of admintools.conf.

Reply:
>Do you think it is necessary to have a step in the operator to ensure admintools.conf is the same on all of the pods?
-- Having a step regularly checking admintools.conf on all Pods may be overkilling, as long as the initiator of AT commands has the latest admintools.conf it should be ok. Maybe we can keep as is at this time, if tests show any problem, then we can see what we can do.

Reply:
I'm not sure about the details of err message, just want to mention that it might be helpful if we know on which nodes the copy succeeded and on which the copy failed, so if there's at least one copy succeeded, then that copy can be the source of latest admintools.conf, such information could be helpful to help debugging and fix admintools.conf

Reply:
@ningdeng based on your comments here, I improved the distributeAdmintoolsConf function.
- the first pod we copy to will always be the same.  It is also the pod that we use for the base for any subsequent admintools.conf changes for install/uninstall
- if the first pod copied was okay, copy to all other pods.  We save the errors and do checking at the end so that we attempt to copy to all pods
- if the copy failed at one of the pods, log an event saying that admintools.conf was partially copied.

You can see these changes in the commit c424386